### ISSUE TITLE
Class pattern matching doesn't narrow `Any`

### ISSUE THREAD

**Issue by thisisrandy** [microsoft/pyright#2674]
**Describe the bug**
When using [class patterns](https://www.python.org/dev/peps/pep-0634/#class-patterns) in `match` statements, the type checker does not appropriately narrow the type of the subject inside `case` blocks.

**To Reproduce**
Open the code below in VSCode and observe characteristics as commented.

**Expected behavior**
Code inside a class pattern-guarded `case` block should be type checked as if the type of the subject has been narrowed as specified in the class pattern.

**Screenshots or Code**
```python
from typing import Any
from dataclasses import dataclass


def matcher(item: Any) -> None:
    match item:
        case str():
            item += 1  # item: Any, no error
        case int():
            item += 1  # item: Never


@dataclass
class Foo:
    bar: str


def matcher_custom(item: Any) -> int:
    match item:
        case Foo():
            return item.bar  # item: Any, bar: Any, no error
        case _:
            return 0
```
**VS Code extension or command-line**
pylance v2021.12.0

**Additional context**
#2628 is possibly related.

**Comment by erictraut** [comment]
PEP 484 and subsequent discussions in the typing community have maintained that `Any` should not be narrowed. There is one exception that most (all?) Python type checkers make to this rule: type narrowing for `isinstance` type guards. (User-defined type guards a'la PEP 637 are also explicitly exempt from this "never narrow Any" rule.)

You're suggesting here that class patterns should be exempted from the "never narrow Any" rule as well. I could maybe be convinced of this, but it's not obvious to me that this is a justifiable exception to the rule. I think we'd want to get the typing community to weigh in on that decision first. None of the other type checkers currently have support for pattern matching, so I can't simply run them to determine what choice they made. Mypy's pattern matching support is the furthest along, but AFAIK it's not yet merged into the master branch.

Out of curiosity, why did you annotate `item` as `Any` rather than making it `object`? If you use `object`, type narrowing works as you expect.

**Comment by thisisrandy** [comment]
I see. It hadn't occured to me that using `object` would have different behavior. That's a very acceptable workaround for the time being.

Perhaps it's just that I don't fully appreciate some subtlety of `Any` that makes narrowing potentially inappropriate. Could you perhaps point to some relevant discussion that might shed some light? As far as my current understanding goes, it seems obvious that narrowing *should* happen. I suspect that, unappreciated subtleties aside, many non-typing experts like myself will expect this behavior.

**Comment by thisisrandy** [comment]
Actually, reading PEP 484 is making this clearer, specifically, "[Any] can be considered a type that has all values and all methods." Still, the PEP also reads, "a value of type Any can be assigned to a variable (or used as a return value) of a more constrained type." I know that't not *literally* what matching is doing, but it feels semantically equivalent. Considering the following, using the `isinstance` exception you cite:

```python
def isinstancer(item: Any) -> None:
    if isinstance(item, str):
        item += 1  # error
    elif isinstance(item, int):
        item += 1  # no error
```

Isn't class matching intended to replace exactly that sort of code with a more readable equivalent?

**Comment by erictraut** [comment]
`Any` effectively tells the type checker that it shouldn't make any assumptions about that type. 

PEP 484 hints at this when it says:
> when a value has type Any, the type checker will allow all operations on it

This topic has come up repeatedly in typeshed, mypy, and typing-sig discussions. I'd be hard pressed to find specific threads. Here's one typing-sig discussion that touches on it: https://mail.python.org/archives/list/typing-sig@python.org/thread/GYVM5KEE6URE6PAH7UTK6324M7GWSFQS/#2AVUELQXVXFXXHIHHLNOBZICUROI2BN2

**Comment by erictraut** [comment]
> Isn't class matching intended to replace exactly that sort of code with a more readable equivalent?
That's the argument I would make in favor of exempting class-based pattern matching from the "never narrow Any" rule. The argument against it is that more exceptions are generally undesirable, and you probably shouldn't be using `Any` to annotate your pattern target expression in the first place.

**Comment by thisisrandy** [comment]
If my vote counts, the explicit equivalence between `if isinstance...` and class-based pattern matching drawn in [PEP 622](https://www.python.org/dev/peps/pep-0622/#rationale-and-goals) justify an exception, because users will emulate the motivating use cases and expect the same result from their type checker, especially if they are refactoring.

That said, possible compromise: My journey to here was predicated on a confusion between `object` and `Any`, or more accurately, not really realizing that `object` was even an option, much less one with the semantics I was hoping for from `Any`. In that light, do you think you could frame this as a warning, e.g. "Any type cannot be [is not] narrowed during class-based pattern matching [and possibly other cases where the user might benefit from a warning]. Consider using the object type instead." I don't know how much work I'm asking for in detecting the appropriate situation(s), but I can say that as I user, I would have benefited.

**Comment by erictraut** [comment]
OK, thanks for the input.

A warning here would be too noisy for most users, so I don't think that's a viable option.

**Comment by thisisrandy** [comment]
Yea, I thought you might say that. In any event, thanks for the clarification and discussion. I look forward to the decision-making on this.

**Comment by erictraut** [comment]
@gvanrossum, as cosponsor of PEP 634, do you have any thoughts on this? Do you think that class pattern matching should be exempt from the "never narrow Any" rule, similar to `isinstance` type guards?

**Comment by erictraut** [comment]
I've also reached out to the Adrian Freund, who is working on the mypy support for pattern matching, to solicit his opinion on this topic. I spent some time looking at his in-progress PR, and I don't see any tests that cover this case, so I'm not sure if he considered it in his implementation.

**Comment by gvanrossum** [comment]
I agree with the OP, `case str()` feels very similar to `if isinstance(item, str)`. In fact PEP 635 explicitly suggests replacing `isinstance()` with a class pattern in the Motivation section.

**Comment by erictraut** [comment]
OK, I've updated the logic for type narrowing related to class patterns so it narrows `Any`, similar to `isinstance`. I also updated the pyright documentation to note this exception to the "never narrow Any" rule.

This change will be in the next release.

**Comment by gvanrossum** [comment]
Great! Maybe you could mention this to the mypy contributor you've reached out to?
________________________________

**Comment by erictraut** [comment]
Yeah, I'm still waiting to hear from him. If he responds, I'll let him know.

**Comment by erictraut** [comment]
This is included in pyright 1.1.196, which I just published. It will also be included in the next release of pylance.
